Add initial P8b smartwatch support

StarGate01 opened the pull request:
This PR extends the existing P8 smartwatch support (eg. from https://github.com/InfiniTimeOrg/InfiniTime/commit/e614af1c4a53e943256b3b2acee4bfe7d7204bbf) to better support more recent models of the P8 series; the P8b. This is a feature desired by the community (see https://github.com/InfiniTimeOrg/InfiniTime/issues/62).

The P8 smartwatches and the Pinetime have a few small differences:
- Slightly different pins are used
- Some variants have a `SC7A20` acceleration sensor, some have a `BMA421` (like the Pinetime)
- Some variants have a `CST716` touch screen sensor, some have a `CST816S` (like the Pinetime)
- Some variants have no external low frequency crystal
- Different SPI flash chips are used (Not relevant as no whitelist is enforced in InfiniTime)

I added support for the `SC7A20` sensor based on the work by @hubmartin and @ildar at https://github.com/ildar/InfiniTime/tree/p8 . A CMake variable (`DRIVER_ACC`) can be used to select which sensor is used at compile time. The acceleration driver code was split into an generic interface definition and two hardware-specific implementations.

In addition, the touch driver now detects which touch sensor chip is used, and adjusts its logic accordingly. The `CST716` does not support an automatic sleep state, and cannot wake from the explicit manual sleep state. On these devices, the `SC7A20` accelerometer is configured to recognize tap and double-tap events. The `CST816S` is not explicitly put to sleep, instead it uses its auto-sleep functionality. On my Pinetime dev kit, I had some issues concerning the initialization of the touch sensor after a reset, so I implemented a small workaround for variants with that issue as well.

It must be said that while this improved touch driver works great on both my P8b watch as well as on my Pinetime dev kit, there are older P8(a) variants which might have incompatible touch sensor configurations. The  `CSTX16` series can be used in two modes - reporting mode and gesture mode - which control when and how often an event is reported to the main CPU. The InfiniTime code currently assumes that both the `CST816S` as well as the `CST716` are configured in reporting mode. The `CST816S` is able to switch at runtime to this mode, which is what the code does. The `CST716` can only be configured once in the factory. In my P8b it came in reporting mode, however there might be other P8 variant with a (yet) incompatible `CST716` in gesture mode. However, I do not have such a watch available, nor have I found a vendor that still sells these (reliably). Adding support for a sensor in gesture mode should be done in a future PR.

The Nimble Bluetooth stack has been slightly modified to be compatible with the LFRC clock source calibration (see https://github.com/apache/mynewt-nimble/issues/1207), as well as the power-conserving `BLE_LL_RFMGMT_ENABLE_TIME` Nimble configuration option (see https://mynewt.apache.org/v1_7_0/network/ble_setup/ble_lp_clock.html#crystal-settle-time-configuration).

A second new CMake option (`LF_CLK`) can be used to specify which source should be used for the low frequency clock (used for Bluetooth). If the `RC` source is chosen, the code will periodically recalibrate the internal RC oscillator using the high frequency clock as a reference. And the third new option (`TARGET_DEVICE`) replaces `WATCH_COLMI_P8` to select the pin map.

All these now options are optional and documented in `buildAndProgram.md`. By default, the regular Pinetime configuration is built. The CMake files were slightly refactored as well, for a better display of the configuration options and better debugging configuration.

A related PR was made to the bootloader: https://github.com/InfiniTimeOrg/pinetime-mcuboot-bootloader/pull/10 .

I realize this is a rather large and broad PR, so please feel free to criticize, and I'll be happy to discuss and incorporate any improvements.

Avamander commented on src/components/settings/Settings.cpp:
```diff
@@ -45,3 +47,14 @@ void Settings::SaveSettingsToFile() {
   fs.FileWrite(&settingsFile, reinterpret_cast<uint8_t*>(&settings), sizeof(settings));
   fs.FileClose(&settingsFile);
 }
+
+void Settings::uploadWakeUpMode() {
+  // Setup the motion sensor interrupt configuration
+  if(isWakeUpModeOn(Settings::WakeUpMode::DoubleTap)) {
+    acc.SetMotion(Drivers::MotionEvents::DoubleTap);
+  } else if(isWakeUpModeOn(Settings::WakeUpMode::SingleTap)) {
+    acc.SetMotion(Drivers::MotionEvents::SingleTap);
+  } else {
+    acc.SetMotion(Drivers::MotionEvents::None);
+  }
+}
```
This seems to be generic wakeup refactoring, must be in a separate PR. Will break other unmerged ones touching similar parts of the code. Very hard to review in this context.

StarGate01 replied:
This controls the gesture that the new accelerometer driver should use to wake up the watch. Useful if the touch sensor cannot generate these events.

Should I break out the accelerometer additions and refactoring (including the new SC7A20 driver) into a new PR?

StarGate01 replied:
This change was postponed to a potential future PR.

Avamander commented on src/main.cpp:
```diff
@@ -302,11 +313,28 @@ void nimble_port_ll_task_func(void* args) {
 
   nrf_drv_clock_init();
 
+  // In some cases (eg. when a watchdog is enabled), 
+  // the NRF SDK incorrectly assumes the LF clock is already running,
+  nrf_drv_clock_lfclk_request(NULL);
+  nrfx_clock_lfclk_start();
+  while(!nrf_clock_lf_is_running()) { }
```
Is this actually a potential issue on the PineTime, should it be wrapped in a preprocessor if?

StarGate01 replied:
These three lines ensure that the clock is running, which is required by the call `nrf_drv_clock_calibration_start` below. 

On the PineTime, no LFRC calibration is used, and the LF clock is started at a later point at `vPortSetupTimerInterrupt` in `FreeRTOS/port_cmsis_systick.c`. The call to `nrf_drv_clock_lfclk_request` is idempotent.

Do you think these three lines should be moved into the `#if(CLOCK_CONFIG_LF_SRC == NRF_CLOCK_LFCLK_RC)` conditional?

StarGate01 replied:
This discussion can be moved to https://github.com/InfiniTimeOrg/InfiniTime/pull/1128 , where this change was broken out into.

StarGate01 commented on src/CMakeLists.txt:
```diff
@@ -924,7 +960,7 @@ add_custom_command(TARGET ${EXECUTABLE_NAME}
 set(EXECUTABLE_MCUBOOT_NAME "pinetime-mcuboot-app")
 set(EXECUTABLE_MCUBOOT_FILE_NAME ${EXECUTABLE_MCUBOOT_NAME}-${pinetime_VERSION_MAJOR}.${pinetime_VERSION_MINOR}.${pinetime_VERSION_PATCH})
-set(IMAGE_MCUBOOT_FILE_NAME ${EXECUTABLE_MCUBOOT_NAME}-image-${pinetime_VERSION_MAJOR}.${pinetime_VERSION_MINOR}.${pinetime_VERSION_PATCH}.bin)
+set(IMAGE_MCUBOOT_FILE_NAME ${EXECUTABLE_MCUBOOT_NAME}-image-${pinetime_VERSION_MAJOR}.${pinetime_VERSION_MINOR}.${pinetime_VERSION_PATCH}.hex)
```
Okay, I will prepare a new PR.

StarGate01 commented:
This PR has been rebased onto https://github.com/InfiniTimeOrg/InfiniTime/pull/1052 , which in turn was rebased onto https://github.com/InfiniTimeOrg/InfiniTime/pull/1051 . Please review / merge those first.

hubmartin commented:
@StarGate01 Thanks for all the awesome work! I'll try that on my P8 soon. I'm glad I could update them after long time.

StarGate01 commented:
@hubmartin Thanks! Let me know how the testing goes, especially if you have a P8a watch (which I can't test on at the moment).

hubmartin commented:
Merged all 3 PRs and on PineTime it works absolutely fine. Doesn't break anything. Correct touch driver is detected.

On P8 also no major issues. Tested and works fine:
- Touch, swipes
- Tap to wake (wow!)
- Wake on Raise Wrist
- Shake wake

Thanks a lot @StarGate01 for your work, my P8 has updated FW after long time.
Don't know which P8 version I have, don't have internal photos, didn't looked at info page before I erased original FW.

https://twitter.com/hubmartin/status/1507463707936370696

![image](https://user-images.githubusercontent.com/497212/160202614-d6878c07-7a2e-49bb-a648-d82b5fd1e8b4.png)

StarGate01 commented:
Great! Thanks for the shoutout :) Your watch appears to be exactly the same as my P8b (see https://github.com/InfiniTimeOrg/InfiniTime/issues/62#issuecomment-846786374). Glad to see the firmware working.

NeroBurner commented on src/drivers/AccelerationSensor.h:
```diff
@@ -0,0 +1,58 @@
+    };
+
+  }
+}
```
please add newline at the end of the file

```suggestion
}

```

NeroBurner commented on src/drivers/Bma421.cpp:
```diff
@@ -126,7 +109,4 @@ void Bma421::SoftReset() {
-Bma421::DeviceTypes Bma421::DeviceType() const {
-  return deviceType;
-}
+}
```
please add newline at the end of the file 

```suggestion
}

```

NeroBurner commented on src/drivers/Bma421.h:
```diff
@@ -27,24 +17,12 @@ namespace Pinetime {
-      bool isOk = false;
-      bool isResetOk = false;
-      DeviceTypes deviceType = DeviceTypes::Unknown;
+
```
I think this newline can be removed as well

```suggestion
```

NeroBurner commented on src/drivers/Cst816s.cpp:
```diff
@@ -21,29 +21,19 @@ Cst816S::Cst816S(TwiMaster& twiMaster, uint8_t twiAddress) : twiMaster {twiMaste
   nrf_gpio_cfg_output(PinMap::Cst816sReset);
   nrf_gpio_pin_clear(PinMap::Cst816sReset);
-  vTaskDelay(5);
+  vTaskDelay(10);
```
does the task delay needs changing? if so please explain why

StarGate01 replied:
The delay of 10 ms is recommended by the datasheet:

> gpio_pin_write(TP_RESET_PIN,0); //Pull down the reset pin
delay_ms(10); // Delay 10ms
gpio_pin_write(TP_RESET_PIN,1); //Pull the reset pin high
delay_ms(50); //It takes more than 30ms initialization time after IC reset to work normally

From a migration guide on CST716 to CST816s. Source: https://download.csdn.net/download/weixin_42680139/27059345 , Download: [cst716与cst816s驱动调试V1.3.zh-CN.en.docx](https://github.com/InfiniTimeOrg/InfiniTime/files/8384468/cst716.cst816s.V1.3.zh-CN.en.docx)

This is probably a leftover from when I was debugging the reset behavior of the chip, see https://github.com/InfiniTimeOrg/InfiniTime/issues/763#issuecomment-1078650363 . If you think this change is unnecessary, I can roll it back no problem. I have not observed a difference (yet).

StarGate01 replied:
The discussion can be continued in https://github.com/InfiniTimeOrg/InfiniTime/pull/1130 .

NeroBurner commented on src/drivers/Cst816s.cpp:
```diff
@@ -21,29 +21,19 @@ Cst816S::Cst816S(TwiMaster& twiMaster, uint8_t twiAddress) : twiMaster {twiMaste
   [0] EnDClick - Enable Double-click action
   */
-  static constexpr uint8_t motionMask = 0b00000101;
+  static constexpr uint8_t motionMask = 0b00000111;
```
is this something specific to the p8b? Otherwise the swipe up and down on the pinetime would be broken? Just curious

StarGate01 replied:
The  register fileds `EnConLR` and `EnConUD` (should) control the "sliding" operation, i.e. event generation while a gesture is performed. LR: Left and right gestures, UD: Up and down gestures. The translation of the field description is odd at best, from what I understand the old code would only enable sliding gestures in the left and right direction. In reality, all directions ware enabled anyways before as well. Either this is a documentation mistake, or a chip behavior issue. Since the canonical way - according to the datasheet - would be to set both bits, I changed it. This might be required for some CST variants. Research is ongoing.

Register map: [CST816.pdf](https://github.com/InfiniTimeOrg/InfiniTime/files/8384517/CST816.pdf) , translation: [CST816S寄存器说明-20190508.zh-CN.en.pdf](https://github.com/InfiniTimeOrg/InfiniTime/files/8384520/CST816S.-20190508.zh-CN.en.pdf)

StarGate01 replied:
The discussion can be continued in https://github.com/InfiniTimeOrg/InfiniTime/pull/1130 .

NeroBurner commented on src/drivers/Cst816s.h:
```diff
@@ -69,6 +77,9 @@ namespace Pinetime {
       uint8_t chipId;
       uint8_t vendorId;
       uint8_t fwVersion;
+      Variant variant;
```
for clairity please initialize the variant variable to unknown explicitly

NeroBurner commented on src/drivers/SC7A20.cpp:
```diff
@@ -0,0 +1,149 @@
+#include <libraries/delay/nrf_delay.h>
+#include "drivers/SC7A20.h"
```
for consistency include the header of this file as first include

NeroBurner commented on src/drivers/SC7A20.h:
```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include "AccelerationSensor.h"
```
please use includes relative to `src`, otherwise I have a hard time maintaing the InfiniSim simulator ;)

NeroBurner commented on src/drivers/SC7A20.h:
```diff
@@ -0,0 +1,31 @@
+      bool isResetOk = false;
+    };
+  }
+}
```
please add a newline at the end of the file

```suggestion
}

```

StarGate01 commented:
I added a new CMake configuration option (`DRIVER_TOUCH`), which is used to control how the CSTx17 touch driver chip is configured. On some models, this configuration is set once in the factory, and cannot be changed or even detected at runtime.
I also released a new build with support for the P8a touch sensor: https://github.com/StarGate01/p8b-infinitime/releases/tag/v1.8.0.2 .

StarGate01 commented:
I have squashed and reordered the commits, and removed the acceleration tap detection functionality, which turned out to be highly unreliable. I hope the commits are now more easily reviewable.

I also edited the PR description above, to reflect the updates to the touch driver compatibility and BLE motion service.

Riksu9000 commented on src/touchhandler/TouchHandler.cpp:
```diff
@@ -50,46 +50,77 @@ Pinetime::Applications::TouchEvents TouchHandler::GestureGet() {
+    }
+  #elif defined(DRIVER_TOUCH_GESTURE) 
+    if (info.gesture != Pinetime::Drivers::Cst816S::Gestures::None) {
+      gesture = ConvertGesture(info.gesture);
+    }
+  #elif defined(DRIVER_TOUCH_DYNAMIC)
+    if (info.gesture != Pinetime::Drivers::Cst816S::Gestures::None) {
+      if (gestureReleased) {
+        if (info.gesture == Pinetime::Drivers::Cst816S::Gestures::SlideDown ||
+            info.gesture == Pinetime::Drivers::Cst816S::Gestures::SlideLeft ||
+            info.gesture == Pinetime::Drivers::Cst816S::Gestures::SlideUp ||
+            info.gesture == Pinetime::Drivers::Cst816S::Gestures::SlideRight ||
+            info.gesture == Pinetime::Drivers::Cst816S::Gestures::LongPress) {
+          if (info.touching) {
+            gesture = ConvertGesture(info.gesture);
+            gestureReleased = false;
+          }
+        } else {
           gesture = ConvertGesture(info.gesture);
-          gestureReleased = false;
         }
-      } else {
-        gesture = ConvertGesture(info.gesture);
       }
     }
-  }
 
-  if (!info.touching) {
-    gestureReleased = true;
-  }
+    if (!info.touching) {
+      gestureReleased = true;
+    }
+  #endif
```
We try to avoid using preprocessor. These different drivers hardly have any shared code, so they should probably be separate classes with a common interface.

StarGate01 replied:
I agree that it would be nice to avoid using preprocessor statements, however I don't know how useful it would actually be to split this piece of code into some kind of abstraction. It is essentially the same driver, only the data it reports is slightly different for each chip.

In any case, we can continue the discussion in https://github.com/InfiniTimeOrg/InfiniTime/pull/1130 .

Riksu9000 reviewed (commented):
Thank you for your work on porting InfiniTIme to other devices, however I think this PR is still too big even with the two other PRs. It would be a good idea to chop it down further to a PR for each driver for example. Also please follow our coding conventions and avoid creating unnecessary changes for a better chance of this being merged.

StarGate01 commented:
@Riksu9000 Thank you for the feedback!

I'll try to split the PR further into chunks, the accelerometer FIFO handling should be easy to separate for example.

Do you prefer to have fewer, larger commits per PR, or more fine-grained ones?

StarGate01 commented on src/displayapp/LittleVgl.cpp:
```diff
@@ -176,17 +176,24 @@ void LittleVgl::FlushDisplay(const lv_area_t* area, lv_color_t* color_p) {
 }
 
-void LittleVgl::SetNewTouchPoint(uint16_t x, uint16_t y, bool contact) {
+void LittleVgl::SetNewTouchPoint(uint16_t x, uint16_t y, bool contact, bool single_event) {
   tap_x = x;
   tap_y = y;
   tapped = contact;
+  tap_single_event = single_event;
 }
 
 bool LittleVgl::GetTouchPadInfo(lv_indev_data_t* ptr) {
   ptr->point.x = tap_x;
   ptr->point.y = tap_y;
   if (tapped) {
     ptr->state = LV_INDEV_STATE_PR;
+    if (tap_single_event) {
+      // If a tap consists of only a single event, enqueue a synthetic release state update
+      tapped = false;
+      tap_single_event = false;
+      return true;
+    }
```
I handled this in https://github.com/InfiniTimeOrg/InfiniTime/pull/1130/commits/2476b34b07b8f834879a25f342127dc9560eb5de , in a new PR at https://github.com/InfiniTimeOrg/InfiniTime/pull/1130 by introducing a new function instead of adding another parameter.

StarGate01 commented on src/drivers/Bma421.cpp:
```diff
@@ -24,96 +22,98 @@ namespace {
-    return;
+  if (ret != BMA4_OK) return;
 
+  // Configure FIFO
+  ret = bma4_set_fifo_config(BMA4_FIFO_ACCEL, 1, &bma);
+  if (ret != BMA4_OK) return;
+  fifo_frame.data = (uint8_t*)&fifo;
+  fifo_frame.length = sizeof(fifo);
+  fifo_frame.fifo_data_enable = BMA4_FIFO_A_ENABLE;
+  fifo_frame.fifo_header_enable = 0;
+
```
Ok, see https://github.com/InfiniTimeOrg/InfiniTime/pull/1132 .

StarGate01 commented:
I have split this PR into seven new ones. I hope this makes the code significantly easier to review.

I listed all the PRs in the stating post of this PR. In the description of each PR dependencies are listed, but most are not depending on each other.

This PR now contains a big merge of all these branches. I'll leave it for historical reasons, or to see all the changes in one place.

Thanks for any help and reviews.

JF002 commented:
@StarGate01 Thank you very much for this extensive work on InfiniTime! These PR are well split and documented :medal_sports: 

Please note that we'll probably need a bit of time to review those PRs as porting to other device is currently not our main priority.

StarGate01 commented:
@JF002 Thanks for taking the time to review. I realize that this is quite a bit of code, so I completely understand that reviewing and adjusting will take its fair time. Any feedback is welcome, I'll do my best to integrate it into the PRs.

ildar commented:
Hello again, pals.
I see many parts of P8 support were accepted and some obviously not.
What are the clue PRs to push yet? I guess https://github.com/InfiniTimeOrg/InfiniTime/pull/1130 ?
Is it much work to do yet? (I've been away for some time, didn't trace all the discussions)
@StarGate01, thanks again and again for your efforts.

StarGate01 commented:
I am planning to rebase my PRs on the most recent Infinitime release quite soon, I recently got a P8 watch again.

Most of the yet unmerged PRs were held up by some discussions about implementation details, which should be able to be resolved eventually.
